File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
import sys
input = sys.stdin.readline
from collections import deque

n, m, v = map(int, input().split(' '))
# 인접리스트 : graph의 인덱스가 각 노드 번호를 말하는 것이고, 해당 노드 별로 연결된 노드를 담는 것 : 메모리효율굿, 속도느림
# 인접행렬 : 노드 i->j로 가는 간선이 있으면 1을 없으면 0을 넣는 것 : 메모리 낭비O, 속도 빠름

# 여기서는 인접리스트를 사용하였고, 앞으로 인접리스트를 사용할 것임
graph = [[] for _ in range(n+1)] # 노드개수 + 1(0포함)
visited = [False]*(n+1) # 노드개수 + 1(0포함)

# 간선의 개수만큼 for문을 돌려야 함 - 노드 간 연결 관계 자체가 간선을 의미
for _ in range(m):
    a, b = map(int, input().split(' '))
    graph[a].append(b)
    graph[b].append(a)
# dfs - 재귀
def dfs(graph, x, visited):
    visited[x] = True
    print(x, end=' ')
    graph[x].sort()
    for i in graph[x]:
        if not visited[i]:
            dfs(graph, i, visited)


# bfs - 큐
def bfs(graph, start, visited):
    queue = deque([start])
    visited[start] = False
    while queue:
        x = queue.popleft()
        print(x, end=' ')
        for i in graph[x]:
            if visited[i]:
                queue.append(i)
                visited[i] = False

dfs(graph, v, visited)
print()
bfs(graph, v, visited)

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.java
import java.util.*;
import java.io.*;

// [[], [], [] ... ] 그래프의 각 인덱스에 연결된 노드 번호를 추가 
// 1번째 인덱스에는 1번과 연결된 노드의 번호 추가!!
// [[], [2,3,4], [1,4], [1,4], [1,2,3]] 

public class Main {
	
	static ArrayList<ArrayList<Integer>> graph;
	static boolean[] visited;
	static Queue<Integer> queue;

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		int n = Integer.parseInt(st.nextToken());
		int m = Integer.parseInt(st.nextToken());
		int v = Integer.parseInt(st.nextToken());
		
		graph = new ArrayList<>(n+1); // 0번 인덱스는 사용X
		for (int i=0; i<=n; i++) {
			graph.add(new ArrayList<>());
		}
		visited = new boolean[n+1];
		queue = new LinkedList<>();
		
		for (int i=0; i<m; i++) {			
			st = new StringTokenizer(br.readLine());
			int node1 = Integer.parseInt(st.nextToken());
			int node2 = Integer.parseInt(st.nextToken());
			
			graph.get(node1).add(node2);
			graph.get(node2).add(node1);
		}
		
		for (ArrayList<Integer> nodes: graph) {
			Collections.sort(nodes);
		}
		
		dfs(v);
		System.out.println();
		bfs(v);
	}
	
	public static void dfs(int x) {
		visited[x] = true;
		System.out.print(x + " ");
		
		for (int i: graph.get(x)) {
			if (!visited[i]) {
				dfs(i);
			}
		}
	}
	
	public static void bfs(int x) {
		queue.offer(x);
		visited[x] = false;
		
		while (!queue.isEmpty()) {
			x = queue.poll();
			System.out.print(x + " ");
			
			for (int i: graph.get(x)) {
				if (visited[i]) {
					queue.offer(i);
					visited[i] = false;
				}
			}
		}
	}
}
